yjit/src/asm/arm64/mod.rs

```diff
@@ -267,7 +267,12 @@ pub fn cmp(cb: &mut CodeBlock, rn: A64Opnd, rm: A64Opnd) {
             DataReg::cmp(rn.reg_no, rm.reg_no, rn.num_bits).into()
         },
         (A64Opnd::Reg(rn), A64Opnd::UImm(imm12)) => {
-            DataImm::cmp(rn.reg_no, imm12.try_into().unwrap(), rn.num_bits).into()
+            let num_bits = if (rn.num_bits == 8) {
+                32 // size can't be 8. assume the remaining 24 bits are zero-extended.
+            } else {
+                rn.num_bits
+            };
```

Comment:
Hmmm. If this instruction doesn't support 8-bit operands then IMO we shouldn't pretend that it does. We should force the user of the instruction to be aware of this.

Reply:
I can see that this is tricky since it seems arm maybe doesn't have 8-bit comparisons? Could we use an `asm.test` here since we are looking at flags and check that the flags are zeroed out above the 8th bit?

Reply:
> Could we use an asm.test here since we are looking at flags and check that the flags are zeroed out above the 8th bit?

I'm not sure if it works. `RUBY_SYMBOL_FLAG` is `0xc` (`.. 0000 1100`), so AND-ing an object with it would check only 3rd and 4th bits. Fixnum could also have those bits, e.g. `INT2FIX(6)` (`(6 << 1) + 1`) is `.. 0000 1101`. 

Assuming `ldurb` + `cmp` is optimal on Arm, we'd have the following options to generate it:

1. The current patch: Write `asm.cmp(obj_opnd.sub_opnd(8), ...)` and let `cmp` on Arm work using 32 bits
   * pros: codegen.rs becomes the cleanest. fewest instructions on both Intel and Arm.
   * cons: confusing as to how Arm actually works
2. Write `asm.cmp(obj_opnd.sub_opnd(8), ...)` and let arm64 passes insert `sub_opnd(32)` or `sub_opnd(64)`.
   * pros: Same generated code, hopefully less magical on Arm
   * cons: The arm64 backend passes will be more complicated?
3. Load `obj_opnd.sub_opnd(8)` and use the loaded value with `sub_opnd(32)` or `sub_opnd(64)`.
   * pros: No magic for Arm
   * cons: One more instruction for Intel

As we don't like option 1 and option 3 is not optimal on Intel, I'm thinking about working on option 2.

Reply:
done https://github.com/ruby/ruby/pull/6359/commits/6d032ca60f31e8bc12a99dfbe504e22cef01af42. I hope that makes sense. It no longer pretends that Arm's `cmp` supports 8-bit operands.